### PR TITLE
[IMP] im_livechat, rating: few improvements regarding ratings

### PR DESCRIPTION
**Pull request by ibo-odoo**
**PURPOSE**

In livechat, rating emojis(happy, neutral and sad) display with percentage of
how happy visitors are with the particular channel, click on that will show all
ratings of livechat channels and stat button is visible while creation, if it
has no rating.

**SPECIFICATION**

Add one computed field "rating_count" for the model 'rating.parent.mixin' which
shows the total rating amount of a particular channel and based on that field
invisible the stat button when the count is 0.
Also, display only happy face emoji with the percentage of happiness on
the kanban view of livechat channel, click on that will only display ratings of
the particular channel.

Task- 2471714

### CODE REVIEW

File: addons/im_livechat/views/im_livechat_channel_views.xml [hunk 1]
```diff
@@ -83,7 +81,7 @@
                             <button class="oe_stat_button" type="action" attrs="{'invisible':[('nbr_channel','=', 0)]}" name="%(mail_channel_action_from_livechat_channel)d" icon="fa-comments">
                                 <field string="Sessions" name="nbr_channel" widget="statinfo"/>
                             </button>
-                            <button name="action_view_rating" attrs="{'invisible':[('rating_percentage_satisfaction','=', -1)]}" class="oe_stat_button" type="object" icon="fa-smile-o">
+                            <button name="action_view_rating" attrs="{'invisible':[('rating_percentage_satisfaction', '&lt;=', 0)]}" class="oe_stat_button" type="object" icon="fa-smile-o">
```
Comment (seb-odoo) [inline addons/im_livechat/views/im_livechat_channel_views.xml:84]: This condition is not correct. Having a rating of 0 is not the same as having no rating.

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Hello,

It seems `rating_percentage_satisfaction` is incorrect with demo data. It says `-1` even though there are sessions that have been rated. Could you take the opportunity to have a look?

File: addons/im_livechat/views/im_livechat_channel_views.xml [hunk 2]
```diff
@@ -51,9 +51,7 @@
                                              <div class="float-right">
                                                 <t t-if="record.rating_percentage_satisfaction.raw_value &gt; 0">
```
Comment (seb-odoo) [inline addons/im_livechat/views/im_livechat_channel_views.xml:52]: Same issue here, this condition is not correct either. It should depend on the number of ratings not their average.

**Comment by ibo-odoo** [comment]
@seb-odoo Done.

**Comment by ibo-odoo** [comment]
@seb-odoo changes are done as per your suggestion. Thanks.!

File: addons/im_livechat/views/im_livechat_channel_views.xml [hunk 3]
```diff
@@ -49,11 +49,9 @@
                                             <i class="fa fa-comments" role="img" aria-label="Comments" title="Comments"></i>
                                             <t t-esc="record.nbr_channel.raw_value"/> Sessions
                                              <div class="float-right">
-                                                <t t-if="record.rating_percentage_satisfaction.raw_value &gt; 0">
+                                                <t t-if="record.rating_percentage_satisfaction.raw_value != -1">
```
Comment (seb-odoo) [inline addons/im_livechat/views/im_livechat_channel_views.xml:52]: You should use the new field here too.

File: addons/rating/models/rating_mixin.py [hunk 4]
```diff
@@ -20,6 +20,15 @@ class RatingParentMixin(models.AbstractModel):
...
         store=False, help="Percentage of happy ratings")
+    rating_count = fields.Integer(
+        "Ratings Count",
+        compute="_compute_rating_count",
+        help="count of ratings for visibility of stat button")
+
+    @api.depends('rating_ids')
+    def _compute_rating_count(self):
+        for record in self:
+            record.rating_count = len(self.rating_ids)
```
Comment (seb-odoo) [inline addons/rating/models/rating_mixin.py:31]: Be careful with this in regard to performance, this requires fetching all of the ratings to count them in python, which might not scale well.

It is probably better to use `search_count` or even `read_group`.

See example in `_compute_rating_stats` in `rating.mixin` itself.

Note that you can probably incorporate the compute into `_compute_rating_percentage_satisfaction` and make a single query for both the count and the percentage (the percentage also uses the count internally, just have to save it into the field).

File: addons/rating/models/rating_mixin.py [hunk 5]
```diff
@@ -20,6 +20,15 @@ class RatingParentMixin(models.AbstractModel):
         "Rating Satisfaction",
         compute="_compute_rating_percentage_satisfaction", compute_sudo=True,
         store=False, help="Percentage of happy ratings")
+    rating_count = fields.Integer(
+        "Ratings Count",
+        compute="_compute_rating_count",
```
Comment (seb-odoo) [inline addons/rating/models/rating_mixin.py:25]: This is probably short enough to fit on a single line.

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Thank you for the update

**Comment by seb-odoo** [comment]
@tde-banana-odoo I'm not sure who manages the `rating` module but I guess it won't hurt if you have a look, is it ok for you to add `rating_count` (non stored) on the parent mixin, similar to what already exist on the rating mixin itself?

**Comment by tde-banana-odoo** [comment]
@seb-odoo I auto-added rating as part of mail framework as it is quite tightly integrated with it. I guess we will work together for this one like with mail: more UI oriented for you, more python oriented for me :D , if it suits you.

File: addons/rating/models/rating_mixin.py [hunk 6]
```diff
@@ -20,6 +20,7 @@ class RatingParentMixin(models.AbstractModel):
         "Rating Satisfaction",
         compute="_compute_rating_percentage_satisfaction", compute_sudo=True,
         store=False, help="Percentage of happy ratings")
+    rating_count = fields.Integer("Ratings Count", compute="_compute_rating_percentage_satisfaction")
```
Comment (tde-banana-odoo) [inline addons/rating/models/rating_mixin.py:23]: I would prefer "# Ratings" as a string, if you don't mind.

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for pinging ! One small remark, otherwise LGTM.

Side remark: could you update a bit commit message ? Some sentences are not easy to understand and commit title does not follow guidelines (should make a valid sentence if concatenated with "if applied, this commit will <title>" .

A small note to explain the new field would be welcomed also.

Thanks !

File: addons/im_livechat/models/im_livechat_channel.py [hunk 7]
```diff
@@ -92,6 +92,16 @@ def action_quit(self):
         self.ensure_one()
         return self.write({'user_ids': [(3, self._uid)]})
 
+    def action_view_channel_ratings(self):
+        self.ensure_one()
+        action = self.env['ir.actions.act_window']._for_xml_id('im_livechat.rating_rating_action_view_livechat_rating')
+        action['display_name'] = _('Ratings for %s') % self.name
```
Comment (seb-odoo) [inline addons/im_livechat/models/im_livechat_channel.py:98]: use named placeholder for translation

File: addons/im_livechat/views/rating_views.xml [hunk 8]
```diff
@@ -26,7 +26,6 @@
         <field name="name">Ratings for livechat channel</field>
         <field name="res_model">rating.rating</field>
         <field name="view_mode">kanban,tree,graph,pivot,form</field>
-        <field name="domain">[('parent_res_model','=','im_livechat.channel'), ('consumed','=',True)]</field>
```
Comment (seb-odoo) [inline addons/im_livechat/views/rating_views.xml:29]: Why did you need to remove this?
Doesn't this make this default action (when not overridden) incorrect?

File: addons/im_livechat/models/im_livechat_channel.py [hunk 9]
```diff
@@ -92,6 +92,16 @@ def action_quit(self):
...
+        action['domain'] = [('parent_res_id', '=', self.id), ('parent_res_model', '=', self._name), ('consumed', '=', True)]
```
Comment (seb-odoo) [inline addons/im_livechat/models/im_livechat_channel.py:99]: `self._name` is not dynamic, it's always `im_livechat.channel` so I would use it explicitly for clarity.

Comment (seb-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:99]: But ideally this should just take the original domain (if you don't remove it) and append the filter on `parent_res_id`.

File: addons/im_livechat/views/rating_views.xml (hunk 8 above)
Comment (seb-odoo) [reply addons/im_livechat/views/rating_views.xml:29]: Note: it would be nice is the stat button was using the new `action_view_channel_ratings` too. Currently it's strange to have a different action open when using the kanban icon or the stat button from form view.

**Review by seb-odoo — CHANGES_REQUESTED** [review]

File: addons/im_livechat/models/im_livechat_channel.py (hunk 7 above)
Comment (ibo-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:98]: Are you suggesting something like `action['display_name'] = _('Ratings for %(channel_name)s', channel_name=self.name)` ?

Comment (seb-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:98]: Yes, this helps translator to know what kind of information they receive.

**Comment by ibo-odoo** [comment]
@seb-odoo Changes are done.. Thanks.!

File: addons/im_livechat/models/im_livechat_channel.py [hunk 10]
```diff
@@ -92,14 +92,14 @@ def action_quit(self):
...
+    def action_view_channel_ratings(self):
         """ Action to display the rating relative to the channel, so all rating of the
             sessions of the current channel
-            :returns : the ir.action 'action_view_rating' with the correct domain
+            :returns : the ir.action 'action_view_channel_ratings' with the correct context
         """
         self.ensure_one()
         action = self.env['ir.actions.act_window']._for_xml_id('im_livechat.rating_rating_action_view_livechat_rating')
-        action['domain'] = [('parent_res_id', '=', self.id), ('parent_res_model', '=', 'im_livechat.channel')]
+        action['context'] = {'search_default_parent_res_name': self.name}
```
Comment (seb-odoo) [inline addons/im_livechat/models/im_livechat_channel.py:102]: Isn't it possible to use id instead of name here?

File: addons/im_livechat/models/im_livechat_channel.py [hunk 11]
```diff
@@ -92,14 +92,14 @@ def action_quit(self):
         self.ensure_one()
         return self.write({'user_ids': [(3, self._uid)]})
 
-    def action_view_rating(self):
+    def action_view_channel_ratings(self):
```
Comment (seb-odoo) [inline addons/im_livechat/models/im_livechat_channel.py:95]: Why do we rename it actually? Wasn't the previous name good enough?

File: addons/im_livechat/views/rating_views.xml [hunk 12]
```diff
@@ -11,7 +11,6 @@
                 <filter string="Code" name="resource" context="{'group_by':'res_name'}"/>
             </xpath>
             <xpath expr="//filter[@name='resource']" position="after">
-                <field name="parent_res_name"/>
```
Comment (seb-odoo) [inline addons/im_livechat/views/rating_views.xml:14]: Do you have a screenshot of the issue when this is not removed?

Comment (ibo-odoo) [reply addons/im_livechat/views/rating_views.xml:14]: Yes..Have a look - https://www.awesomescreenshot.com/image/7025723?key=acdf88d8f8b6ced48d546873d386c72f

File: addons/im_livechat/models/im_livechat_channel.py (hunk 10 above)
Comment (ibo-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:102]: No filteration is not working with `id`. I tried.

**Comment by ibo-odoo** [comment]
Hello.. @seb-odoo the branch is rebased .Thanks.!

File: addons/im_livechat/models/im_livechat_channel.py (hunk 10 above)
Comment (seb-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:102]: Hello,

You need to add the field in the search view of `rating.rating` to be able to use it with a `search_default_`.

`<field name="parent_res_id" filter_domain="[('parent_res_id','ilike',self)]"/>`

Then it's possible to use id here, which is more robust than name.

Comment (ibo-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:102]: Hello,
As per your this suggestion, I tried and yes it does filteration with ID.

But we need to do same as project rating do as per @LNA's suggestion. In project, default search is done with 'parent document name(name)'(https://drive.google.com/file/d/1GoP_Cb-7xSE44ZyJdkEGB0EiSI4JY6ZS/view?usp=sharing) and as we did filteration with id here so it shows 'parent document(id)'(https://drive.google.com/file/d/1nLW6GhJt9clorlIEoE5_QfjfxPP0YE7G/view?usp=sharing) in header thus its not the same as project rating. May be we can change the string(parent document) but can't change the value(Id).

Comment (seb-odoo) [reply addons/im_livechat/models/im_livechat_channel.py]: I will ask him on the task, thanks for the feedback.

Comment (ibo-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:101]: Hello @seb-odoo 

As per CPE's opinion on the task (`parent document name(channel name) is more understandable when filtering the ratings`
Updated commit accordingly. Thanks.

Comment (seb-odoo) [reply addons/im_livechat/models/im_livechat_channel.py:101]: Ok, thank you for considering it and for the update.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+

**Comment by seb-odoo** [comment]
@robodoo r+
